### ISSUE TITLE
"Discard line" in git diff panel does not work

### ISSUE THREAD

**Issue by daattali** [rstudio/rstudio#12635]
I've noticed this for years but was always too lazy to report it.

I'm currently on Windows 10, RStudio build 576

When viewing a git diff in RStudio's git viewer, each changed line has "Stage line" and "Discard line" buttons. See below. Clicking the button never does anything.

![image](https://user-images.githubusercontent.com/952340/214789454-973ad098-463e-4ae1-87c4-6f1f28a77d9f.png)

**Comment by ronblum** [comment]
@daattali Thank you for raising this! I can reproduce this on MacOS 13.2 using RStudio Desktop 2022.07.2+576 and 2022.12.0+353. It's only for the last change—the rest work. Do you see the same behavior?

**Comment by daattali** [comment]
What do you mean "only for the last change "?

**Comment by ronblum** [comment]
I mean that if you have multiple changes, the buttons work for all of the changes except for the last one. But now I can't reproduce that scenario.

**Comment by kevinushey** [comment]
You can run `Sys.setenv(RSTUDIO_GIT_LOG = 1)` if you want to see what git commands RStudio is trying to execute behind the scenes. Presumedly we're synthesizing the wrong command, or something else is going wrong.

**Comment by daattali** [comment]
Here is the output I see from one click on "Discard line"

```r
>>> C:/Program Files/Git/bin/git.exe apply -- C:/Users/Dean-X1C/AppData/Local/Temp/RtmpqQLQiZ/rstudiovcs75a02eba5cbb.patch
>>> C:/Program Files/Git/bin/git.exe status -z --porcelain -- C:/Users/Dean-X1C/Documents/R/ggExtra
>>> C:/Program Files/Git/bin/git.exe branch --no-color -a
>>> C:/Program Files/Git/bin/git.exe branch --no-color -a
>>> C:/Program Files/Git/bin/git.exe config --get branch.master.remote
>>> C:/Program Files/Git/bin/git.exe config --get branch.master.merge
>>> C:/Program Files/Git/bin/git.exe log @{u}..HEAD --oneline
>>> C:/Program Files/Git/bin/git.exe status -z --porcelain -- C:/Users/Dean-X1C/Documents/R/ggExtra
>>> C:/Program Files/Git/bin/git.exe branch --no-color -a
>>> C:/Program Files/Git/bin/git.exe branch --no-color -a
>>> C:/Program Files/Git/bin/git.exe config --get branch.master.remote
>>> C:/Program Files/Git/bin/git.exe config --get branch.master.merge
>>> C:/Program Files/Git/bin/git.exe log @{u}..HEAD --oneline
>>> C:/Program Files/Git/bin/git.exe diff -U5 -w -- C:/Users/Dean-X1C/Documents/R/ggExtra/R/ggMarginalGadget.R
```

And here's a GIF showing me trying to click it on both the green and red diffed lines. At the end I show that "Discard chunk" does work but not "Discard line"

![shiny](https://user-images.githubusercontent.com/952340/215874063-2f41f4cd-fbc9-4795-9175-9aacf08580ee.gif)

**Comment by kevinushey** [comment]
> >>> C:/Program Files/Git/bin/git.exe apply -- C:/Users/Dean-X1C/AppData/Local/Temp/RtmpqQLQiZ/rstudiovcs75a02eba5cbb.patch

Unfortunate that we don't get to see the patch file that was created...

Do you see anything in the request log? (Help -> Diagnostics -> Request Log). There should be an event like `git_apply_patch`, with the parameters we computed for the intended patch. It might help to see those.

It may also be useful to know if you see any errors in the logs; we should be logging something if we were applying an incorrect patch.

**Comment by daattali** [comment]
I can see the requests in the diagnostics, but no errors. Here's the first:

```r
Request ID: -1642625296

== REQUEST ======
{
  "method": "git_apply_patch",
  "params": [
    "--- a/R/ggMarginalGadget.R\n+++ b/R/ggMarginalGadget.R\n@@ -17,21 +17,22 @@\n #'   plot <- ggplot2::ggplot(mtcars, ggplot2::aes(wt, mpg)) + ggplot2::geom_point()\n #'   plot2 <- ggMarginalGadget(plot)\n #' }\n ggMarginalGadget <- function(plot) {\n   if (missing(plot)) {\n     stop(\"You must provide a ggplot2 plot.1\", call. = FALSE)\n   }\n \n   ggMarginalGadgetHelper(deparse(substitute(plot)), addin = FALSE)\n }\n \n ggMarginalGadgetAddin <- function() {\n   context <- rstudioapi::getActiveDocumentContext()\n   text <- context$selection[[1]]$text\n   if (nchar(text) == 0) {\n+    stop(\"Please highlight a ggplot2 plot before selecting this addin.\")\n     stop(\"Please highlight a ggplot2 plot before selecting this addin2.\")\n   }\n   ggMarginalGadgetHelper(text, addin = TRUE)\n }\n \n #' @import shiny\n@@ -78,11 +79,11 @@\n \n   if (!ggplot2::is.ggplot(get(plotname, envir = .GlobalEnv))) {\n     stop(\"You did not provide a ggplot2 plot.\", call. = FALSE)\n   }\n \n   resourcePath <- system.file(\"gadgets\", \"ggmargi3nal\", package = \"ggExtra\")\n   shiny::addResourcePath(\"ggm\", resourcePath)\n \n   ui <- miniPage(\n     shinyjs::useShinyjs(),\n     tags$head(includeCSS(file.path(resourcePath, \"css\", \"app.css\"))),\n",
    0,
    "UTF-8"
  ],
  "sourceWnd": "review_changes",
  "clientId": "33e600bb-c1b1-46bf-b562-ab5cba070b0e",
  "clientVersion": ""
}

== RESPONSE ======
{
  "result": null
}
```

**Comment by kevinushey** [comment]
Thanks -- so evidently this is the patch it was trying to apply in that case.

```
--- a/R/ggMarginalGadget.R
+++ b/R/ggMarginalGadget.R
@@ -17,21 +17,22 @@
 #'   plot <- ggplot2::ggplot(mtcars, ggplot2::aes(wt, mpg)) + ggplot2::geom_point()
 #'   plot2 <- ggMarginalGadget(plot)
 #' }
 ggMarginalGadget <- function(plot) {
   if (missing(plot)) {
     stop(\"You must provide a ggplot2 plot.1\", call. = FALSE)
   }
 
   ggMarginalGadgetHelper(deparse(substitute(plot)), addin = FALSE)
 }
 
 ggMarginalGadgetAddin <- function() {
   context <- rstudioapi::getActiveDocumentContext()
   text <- context$selection[[1]]$text
   if (nchar(text) == 0) {
+    stop(\"Please highlight a ggplot2 plot before selecting this addin.\")
     stop(\"Please highlight a ggplot2 plot before selecting this addin2.\")
   }
   ggMarginalGadgetHelper(text, addin = TRUE)
 }
 
 #' @import shiny
@@ -78,11 +79,11 @@
 
   if (!ggplot2::is.ggplot(get(plotname, envir = .GlobalEnv))) {
     stop(\"You did not provide a ggplot2 plot.\", call. = FALSE)
   }
 
   resourcePath <- system.file(\"gadgets\", \"ggmargi3nal\", package = \"ggExtra\")
   shiny::addResourcePath(\"ggm\", resourcePath)
 
   ui <- miniPage(
     shinyjs::useShinyjs(),
     tags$head(includeCSS(file.path(resourcePath, \"css\", \"app.css\"))),
```

**Comment by nathanckuhn** [comment]
Chiming in to say that I'm seeing this same bug today. I'm on Windows 10 Enterprise & running RStudio 2022.12.0 Build 353.
The "Stage line" & "Discard line" buttons DO work when there's only one chunk, but as soon as there are multiple chunks neither the "Stage line" or "Discard line" buttons do anything. Interestingly, the "Stage chunk" & "Discard chunk" buttons seem to work fine with one or multiple chunks.
(I don't check GitHub often, so I might not reply quickly -- just wanted to confirm I'm seeing it too! Thanks to whoever is working on this!)

**Comment by daattali** [comment]
bad bot. Issue is still happening

**Comment by olivroy** [comment]
This appears to be fixed in the latest daily !! #5476
